libminifi/src/c2/C2Agent.cpp

```diff
@@ -75,54 +78,55 @@ C2Agent::C2Agent(const std::shared_ptr<core::controller::ControllerServiceProvid
 
   c2_producer_ = [&]() {
     // place priority on messages to send to the c2 server
-      if (protocol_.load() != nullptr && request_mutex.try_lock_for(std::chrono::seconds(1))) {
-        std::lock_guard<std::timed_mutex> lock(request_mutex, std::adopt_lock);
-        if (!requests.empty()) {
-          int count = 0;
-          do {
-            const C2Payload payload(std::move(requests.back()));
-            requests.pop_back();
-            try {
-              C2Payload && response = protocol_.load()->consumePayload(payload);
-              enqueue_c2_server_response(std::move(response));
-            }
-            catch(const std::exception &e) {
-              logger_->log_error("Exception occurred while consuming payload. error: %s", e.what());
-            }
-            catch(...) {
-              logger_->log_error("Unknonwn exception occurred while consuming payload.");
-            }
-          }while(!requests.empty() && ++count < max_c2_responses);
+    if (protocol_.load() != nullptr) {
+      std::vector<C2Payload> payload_batch;
+      payload_batch.reserve(max_c2_responses);
+      auto getRequestPayload = [&payload_batch] (C2Payload&& payload) { payload_batch.emplace_back(std::move(payload)); };
+      const std::chrono::system_clock::time_point timeout_point = std::chrono::system_clock::now() + std::chrono::milliseconds(1);
+      for (std::size_t attempt_num = 0; attempt_num < max_c2_responses; ++attempt_num) {
+        if (!requests.consumeWaitUntil(getRequestPayload, timeout_point)) {
```

Comment:
Previously we took the immediately available payloads only. Now we wait 1ms. Why is this change?

If we reverted to the old way, it would also no longer be a problem to reunify the first two loops and eliminate the need for the temporary buffer (`payload_batch`).

Reply:
Good catch, we want to wait 1 second, not 1 millisecond on locking.

Reply:
The mutex inside the queue is not `timed_mutex` so we wait indefinitely for locking and only wait definitely for content.

Reply:
What is the recommendation here?

Reply:
I understand that the indefinite waiting on the mutex is because of the use of `MinifiConcurrentQueue`. But why did we introduce a waiting on the queue condition variable for elements?

I just want an explanation, to see that the change was intentional and why.

Reply:
I just thought that if one second waiting time was reasonable here, than I should keep this if-it-works-in-one-second-it-is-good behaviour. Maybe there should be a timeout on the mutex itself, but I wonder what solution sounds reasonable here?

Reply:
In the old ideal (and likely common) case before, we got the mutex instantly, took whatever was in the queue and continued execution.

In the new common case, we lock the mutex, take whatever is in the queue, wait 1 second for more to arrive and continue execution.

The common case execution time is increased by almost 1 second, which is probably orders of magnitude more than how much time is spent executing business logic. I'd like to get rid of that waiting if it's not needed. I suggest using `consume` from `ConcurrentQueue` without waiting on the condition variable. This way if we assume little waiting for the mutex, the execution time stays similar.